File: football/Models/TeamAbbreviations.swift
import Foundation

class TeamAbbreviations {
    static let shared = TeamAbbreviations()

    // 3글자 약어
    private let abbreviations: [String: String] = [
        // Premier League
        "Arsenal": "ARS",
        "Aston Villa": "AVL",
        "Bournemouth": "BOU",
        "Brentford": "BRE",
        "Brighton": "BHA",
        "Burnley": "BUR",
        "Chelsea": "CHE",
        "Crystal Palace": "CRY",
        "Everton": "EVE",
        "Fulham": "FUL",
        "Liverpool": "LIV",
        "Luton Town": "LUT",
        "Manchester City": "MCI",
        "Manchester United": "MUN",
        "Newcastle United": "NEW",
        "Nottingham Forest": "NFO",
        "Sheffield United": "SHU",
        "Tottenham Hotspur": "TOT",
        "West Ham United": "WHU",
        "Wolverhampton Wanderers": "WOL",
        // La Liga
        "Real Madrid": "RMA",
        "Barcelona": "BAR",
        "Atletico Madrid": "ATM",
        "Sevilla": "SEV",
        "Real Sociedad": "RSO",
        "Real Betis": "BET",
        "Villarreal": "VIL",
        "Valencia": "VAL",
        "Athletic Club": "ATH",
        "Getafe": "GET",
        "Osasuna": "OSA",
        "Celta Vigo": "CEL",
        "Almeria": "ALM",
        "Cadiz": "CAD",
        "Granada": "GRA",
        "Mallorca": "MLL",
        "Las Palmas": "LPA",
        "Girona": "GIR",
        // Serie A
        "Inter Milan": "INT",
        "AC Milan": "MIL",
        "Juventus": "JUV",
        "Napoli": "NAP",
        "Roma": "ROM",
        "Lazio": "LAZ",
        "Atalanta": "ATA",
        "Torino": "TOR",
        "Fiorentina": "FIO",
        "Bologna": "BOL",
        "Udinese": "UDI",
        "Sassuolo": "SAS",
        "Empoli": "EMP",
        "Genoa": "GEN",
        "Lecce": "LEC",
        "Cagliari": "CAG",
        "Monza": "MON",
        "Salernitana": "SAL",
        "Hellas Verona": "VER",
        // Bundesliga
        "Bayern Munich": "FCB",
        "Bayern München": "FCB",
        "Borussia Dortmund": "BVB",
        "RB Leipzig": "RBL",
        "Bayer Leverkusen": "LEV",
        "Union Berlin": "UNB",
        "Freiburg": "SCF",
        "Eintracht Frankfurt": "SGE",
        "Wolfsburg": "WOB",
        "Mainz": "M05",
        "Borussia Mönchengladbach": "BMG",
        "Cologne": "KOE",
        "Augsburg": "FCA",
        "Stuttgart": "VFB",
        "Werder Bremen": "SVW",
        "Bochum": "BOC",
        "Heidenheim": "FCH",
        "Darmstadt": "SVD",
        "Hoffenheim": "TSG",
        "1.FC Heidenheim": "FCH",
        "1. FC Heidenheim": "FCH",
        "1 FC Heidenheim": "FCH",
        "FC Heidenheim": "FCH",
        // Ligue 1
        "Paris Saint-Germain": "PSG",
        "Paris Saint Germain": "PSG",
        "Marseille": "OM",
        "Lyon": "LYO",
        "Monaco": "ASM",
        "Lille": "LIL",
        "Rennes": "REN",
        "Nice": "OGC",
        "Lens": "LEN",
        "Toulouse": "TOU",
        "Nantes": "FCN",
        "Strasbourg": "STR",
        "Montpellier": "MON",
        "Brest": "BRE",
        "Reims": "REI",
        "Metz": "MET",
        "Clermont": "CLE",
        "Le Havre": "HAC",
        "Lorient": "LOR",

        // European Competitions (UCL & UEL Regulars)
        "Benfica": "BEN",
        "Porto": "POR",
        "Sporting CP": "SCP",
        "Celtic": "CEL",
        "Rangers": "RAN",
        "Shakhtar Donetsk": "SHA",
        "Dynamo Kyiv": "DYN",
        "Red Star Belgrade": "RSB",
        "Olympiacos": "OLY",
        "Galatasaray": "GAL",
        "Fenerbahce": "FEN",
        "Besiktas": "BES",
        "Ajax": "AJA",
        "Feyenoord": "FEY",
        "PSV": "PSV",
        "Club Brugge": "CBR",
        "Anderlecht": "AND",
        "Basel": "BAS",
        "Young Boys": "YBO",
        "Dinamo Zagreb": "DZG",
        "Sheriff Tiraspol": "SHF",
        "Slavia Prague": "SLP",
        "Sparta Prague": "SPA",
        "Ludogorets": "LUD",
        "Ferencvaros": "FER"
    ]

    // 줄임 이름 (Shortened Names)
    private let shortenedNames: [String: String] = [
        // Premier League
        "Manchester United": "Man United",
        "Manchester City": "Man City",
        "Tottenham Hotspur": "Tottenham",
        "Newcastle United": "Newcastle",
        "Nottingham Forest": "Nott'm Forest",
        "Sheffield United": "Sheffield Utd",
        "West Ham United": "West Ham",
        "Wolverhampton Wanderers": "Wolves",
        "Brighton & Hove Albion": "Brighton",
        "Brighton and Hove Albion": "Brighton",
        "Crystal Palace": "C. Palace",
        
        // La Liga
        // "Atletico Madrid": "Atlético", // 원래 이름 그대로 표시
        // "Athletic Club": "Athletic", // 원래 이름 그대로 표시
        "Real Sociedad": "R. Sociedad",
        "Real Betis": "Betis",
        "Real Valladolid": "Valladolid",
        "Celta Vigo": "Celta",
        "Espanyol Barcelona": "Espanyol",
        "Deportivo Alaves": "Alavés",
        "Rayo Vallecano": "Rayo",
        
        // Bundesliga
        // "Bayern Munich": "Bayern", // 원래 이름 그대로 표시
        // "Bayern München": "Bayern", // 원래 이름 그대로 표시
        "Borussia Dortmund": "Dortmund",
        "Borussia Mönchengladbach": "M'gladbach",
        "Bayer Leverkusen": "Leverkusen",
        "RB Leipzig": "Leipzig",
        "Eintracht Frankfurt": "Frankfurt",
        "Union Berlin": "Union",
        "Werder Bremen": "Bremen",
        "1.FC Heidenheim": "Heidenheim",
        "1. FC Heidenheim": "Heidenheim",
        "1 FC Heidenheim": "Heidenheim",
        "FC Heidenheim": "Heidenheim",
        "1899 Hoffenheim": "Hoffenheim",
        "TSG 1899 Hoffenheim": "Hoffenheim",
        "TSG Hoffenheim": "Hoffenheim",
        "FSV Mainz 05": "Mainz",
        "1. FSV Mainz 05": "Mainz",
        "1.FSV Mainz 05": "Mainz",
        "1 FSV Mainz 05": "Mainz",
        "VfL Wolfsburg": "Wolfsburg",
        "VFL Wolfsburg": "Wolfsburg",
        "SV Elversberg": "Elversberg",
        "SV 07 Elversberg": "Elversberg",
        "1. FC Köln": "Köln",
        "1.FC Köln": "Köln",
        "1 FC Köln": "Köln",
        "FC Köln": "Köln",
        "FC St. Pauli": "St. Pauli",
        "FC St Pauli": "St. Pauli",
        "St. Pauli": "St. Pauli",
        "FC Augsburg": "Augsburg",
        "VfB Stuttgart": "Stuttgart",
        "VfL Bochum": "Bochum",
        "VFL Bochum": "Bochum",
        "SC Freiburg": "Freiburg",
        "SV Darmstadt 98": "Darmstadt",
        "SV Darmstadt": "Darmstadt",
        "Fortuna Düsseldorf": "Düsseldorf",
        "Fortuna Dusseldorf": "Düsseldorf",
        "Holstein Kiel": "Kiel",
        
        // Ligue 1
        "Paris Saint-Germain": "PSG",
        "Paris Saint Germain": "PSG",
        "Olympique Lyonnais": "Lyon",
        "Olympique Marseille": "Marseille",
        "Olympique de Marseille": "Marseille",
        
        // Serie A
        "Inter Milan": "Inter",
        "AC Milan": "Milan",
        "Hellas Verona": "Verona",
        "Internazionale Milano": "Inter"
    ]

    private init() {}

    // 3글자 약어 반환
    static func abbreviation(for teamName: String) -> String {
        let normalized = teamName
            .replacingOccurrences(of: "-", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return shared.abbreviations[normalized] ?? String(teamName.prefix(3)).uppercased()
    }
    
    // 줄임 이름 반환
    static func shortenedName(for teamName: String) -> String {
        let normalized = teamName
            .replacingOccurrences(of: "-", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return shared.shortenedNames[normalized] ?? teamName
    }
}

File: web/lib/utils/team-abbreviations.ts
// 팀명 약어 매핑
const teamAbbreviations: Record<string, string> = {
  // Premier League
  'Manchester United': 'MUN',
  'Manchester City': 'MCI',
  'Liverpool': 'LIV',
  'Chelsea': 'CHE',
  'Arsenal': 'ARS',
  'Tottenham': 'TOT',
  'Tottenham Hotspur': 'TOT',
  'Newcastle': 'NEW',
  'Newcastle United': 'NEW',
  'Leicester': 'LEI',
  'Leicester City': 'LEI',
  'West Ham': 'WHU',
  'West Ham United': 'WHU',
  'Everton': 'EVE',
  'Aston Villa': 'AVL',
  'Crystal Palace': 'CRY',
  'Brighton': 'BHA',
  'Brighton & Hove Albion': 'BHA',
  'Fulham': 'FUL',
  'Brentford': 'BRE',
  'Nottingham Forest': 'NFO',
  'Bournemouth': 'BOU',
  'Wolverhampton': 'WOL',
  'Wolverhampton Wanderers': 'WOL',
  'Wolves': 'WOL',
  'Burnley': 'BUR',
  'Sheffield United': 'SHU',
  'Luton': 'LUT',
  'Luton Town': 'LUT',
  'Southampton': 'SOU',
  'Leeds': 'LEE',
  'Leeds United': 'LEE',
  'Ipswich': 'IPS',
  'Ipswich Town': 'IPS',
  
  // La Liga
  'Real Madrid': 'RMA',
  'Barcelona': 'BAR',
  'Atletico Madrid': 'ATM',
  'Atlético Madrid': 'ATM',
  'Real Sociedad': 'RSO',
  'Real Betis': 'BET',
  'Sevilla': 'SEV',
  'Villarreal': 'VIL',
  'Athletic Bilbao': 'ATH',
  'Athletic Club': 'ATH',
  'Valencia': 'VAL',
  'Osasuna': 'OSA',
  'Getafe': 'GET',
  'Girona': 'GIR',
  'Rayo Vallecano': 'RAY',
  'Celta Vigo': 'CEL',
  'Mallorca': 'MAL',
  'Cadiz': 'CAD',
  'Cádiz': 'CAD',
  'Alaves': 'ALA',
  'Alavés': 'ALA',
  'Granada': 'GRA',
  'Las Palmas': 'LPA',
  'Almeria': 'ALM',
  'Almería': 'ALM',
  'Real Valladolid': 'VLD',
  'Espanyol': 'ESP',
  'Elche': 'ELC',
  
  // Serie A
  'Juventus': 'JUV',
  'AC Milan': 'MIL',
  'Inter': 'INT',
  'Inter Milan': 'INT',
  'Napoli': 'NAP',
  'Roma': 'ROM',
  'AS Roma': 'ROM',
  'Lazio': 'LAZ',
  'Atalanta': 'ATA',
  'Fiorentina': 'FIO',
  'Torino': 'TOR',
  'Bologna': 'BOL',
  'Udinese': 'UDI',
  'Sassuolo': 'SAS',
  'Monza': 'MON',
  'Empoli': 'EMP',
  'Salernitana': 'SAL',
  'Lecce': 'LEC',
  'Verona': 'VER',
  'Hellas Verona': 'VER',
  'Spezia': 'SPE',
  'Cremonese': 'CRE',
  'Sampdoria': 'SAM',
  'Cagliari': 'CAG',
  'Genoa': 'GEN',
  'Frosinone': 'FRO',
  
  // Bundesliga
  'Bayern Munich': 'BAY',
  'Bayern München': 'BAY',
  'Borussia Dortmund': 'BVB',
  'Dortmund': 'BVB',
  'RB Leipzig': 'RBL',
  'Leipzig': 'RBL',
  'Bayer Leverkusen': 'B04',
  'Leverkusen': 'B04',
  'Union Berlin': 'UNI',
  'Eintracht Frankfurt': 'SGE',
  'Frankfurt': 'SGE',
  'Freiburg': 'FRE',
  'SC Freiburg': 'FRE',
  'Hoffenheim': 'TSG',
  '1899 Hoffenheim': 'TSG',
  'Wolfsburg': 'WOB',
  'VfL Wolfsburg': 'WOB',
  'Borussia Monchengladbach': 'BMG',
  'Borussia Mönchengladbach': 'BMG',
  'Monchengladbach': 'BMG',
  'Mainz': 'M05',
  'Mainz 05': 'M05',
  'Augsburg': 'AUG',
  'FC Augsburg': 'AUG',
  'Hertha Berlin': 'BSC',
  'Hertha BSC': 'BSC',
  'VfB Stuttgart': 'VFB',
  'Stuttgart': 'VFB',
  'Schalke': 'S04',
  'Schalke 04': 'S04',
  'Werder Bremen': 'SVW',
  'Bremen': 'SVW',
  'Bochum': 'BOC',
  'VfL Bochum': 'BOC',
  'FC Koln': 'KOL',
  'FC Köln': 'KOL',
  'Cologne': 'KOL',
  'Heidenheim': 'HDH',
  'Darmstadt': 'D98',
  'Darmstadt 98': 'D98',
  
  // Ligue 1
  'Paris Saint Germain': 'PSG',
  'Paris Saint-Germain': 'PSG',
  'PSG': 'PSG',
  'Marseille': 'MAR',
  'Olympique Marseille': 'MAR',
  'Monaco': 'MON',
  'AS Monaco': 'MON',
  'Lyon': 'OL',
  'Olympique Lyon': 'OL',
  'Olympique Lyonnais': 'OL',
  'Lille': 'LIL',
  'Nice': 'NIC',
  'OGC Nice': 'NIC',
  'Rennes': 'REN',
  'Stade Rennes': 'REN',
  'Lens': 'LEN',
  'RC Lens': 'LEN',
  'Toulouse': 'TOU',
  'Montpellier': 'MTP',
  'Strasbourg': 'STR',
  'RC Strasbourg': 'STR',
  'Nantes': 'NAN',
  'FC Nantes': 'NAN',
  'Reims': 'REI',
  'Stade Reims': 'REI',
  'Brest': 'BRE',
  'Stade Brest': 'BRE',
  'Lorient': 'LOR',
  'FC Lorient': 'LOR',
  'Clermont': 'CLE',
  'Clermont Foot': 'CLE',
  'Metz': 'MET',
  'FC Metz': 'MET',
  'Le Havre': 'HAC',
  'Le Havre AC': 'HAC',
  'Auxerre': 'AUX',
  'AJ Auxerre': 'AUX',
  'Angers': 'ANG',
  'Angers SCO': 'ANG',
  'Troyes': 'TRO',
  'Saint-Etienne': 'STE',
  'Saint-Étienne': 'STE',
  'AS Saint-Étienne': 'STE',
  'Bordeaux': 'BOR',
  
  // Champions League / International
  'Real Madrid CF': 'RMA',
  'FC Barcelona': 'BAR',
  'Liverpool FC': 'LIV',
  'Manchester United FC': 'MUN',
  'Chelsea FC': 'CHE',
  'Arsenal FC': 'ARS',
  'Manchester City FC': 'MCI',
  'Tottenham Hotspur FC': 'TOT',
  'FC Bayern München': 'BAY',
  'Juventus FC': 'JUV',
  'FC Inter Milan': 'INT',
  'SSC Napoli': 'NAP',
  'Atalanta BC': 'ATA',
  'SS Lazio': 'LAZ',
  'Bayer 04 Leverkusen': 'B04',
  
  // 기타 유럽팀들
  'Ajax': 'AJA',
  'PSV': 'PSV',
  'PSV Eindhoven': 'PSV',
  'Feyenoord': 'FEY',
  'Celtic': 'CEL',
  'Rangers': 'RAN',
  'Benfica': 'BEN',
  'SL Benfica': 'BEN',
  'Porto': 'POR',
  'FC Porto': 'POR',
  'Sporting CP': 'SCP',
  'Sporting': 'SCP',
  'Club Brugge': 'BRU',
  'Club Bruges': 'BRU',
  'Anderlecht': 'AND',
  'Galatasaray': 'GAL',
  'Fenerbahce': 'FEN',
  'Fenerbahçe': 'FEN',
  'Besiktas': 'BES',
  'Beşiktaş': 'BES',
  'Shakhtar Donetsk': 'SHA',
  'Dynamo Kyiv': 'DYK',
  'Red Bull Salzburg': 'RBS',
  'FC Salzburg': 'RBS',
  'Young Boys': 'YB',
  'BSC Young Boys': 'YB',
  'FC Basel': 'BAS',
  'Basel': 'BAS',
  'FC Copenhagen': 'COP',
  'Copenhagen': 'COP',
  'Malmo': 'MAL',
  'Malmö FF': 'MAL',
  'Red Star Belgrade': 'RSB',
  'Crvena Zvezda': 'RSB',
  'PAOK': 'PAO',
  'Olympiacos': 'OLY',
  'AEK Athens': 'AEK',
}

/**
 * 팀명을 약어로 변환
 * @param teamName 원본 팀명
 * @returns 3글자 약어 또는 원본 팀명의 처음 3글자
 */
export function getTeamAbbreviation(teamName: string): string {
  // 매핑에 있는 경우 약어 반환
  if (teamAbbreviations[teamName]) {
    return teamAbbreviations[teamName]
  }
  
  // FC, CF, SC 등의 접미사 제거
  const cleanName = teamName
    .replace(/\s?(FC|CF|SC|AC|AS|SS|SSC|BC|BSC|VfL|VfB|TSV|SV|KV|RCD|UD|SD|CD|AD)\s?/gi, '')
    .trim()
  
  // 정리된 이름으로 다시 검색
  if (teamAbbreviations[cleanName]) {
    return teamAbbreviations[cleanName]
  }
  
  // 매핑에 없는 경우
  // 1. 공백으로 분리하여 각 단어의 첫 글자 추출
  const words = teamName.split(/\s+/)
  if (words.length >= 2) {
    // 최대 3글자까지만
    return words
      .map(word => word[0])
      .join('')
      .toUpperCase()
      .slice(0, 3)
  }
  
  // 2. 단일 단어인 경우 처음 3글자
  return teamName.slice(0, 3).toUpperCase()
}

/**
 * 모바일 여부에 따라 팀명 또는 약어 반환
 * @param teamName 원본 팀명
 * @param isMobile 모바일 여부
 * @returns 팀명 또는 약어
 */
export function getTeamDisplayName(teamName: string, isMobile: boolean): string {
  return isMobile ? getTeamAbbreviation(teamName) : teamName
}